src/test/java/bio/terra/workspace/db/ResourceDaoTest.java

```diff
@@ -59,5 +63,87 @@ public void createGetControlledAiNotebookInstance() {
 
     assertEquals(
         resource, resourceDao.getResource(resource.getWorkspaceId(), resource.getResourceId()));
+
+    resourceDao.deleteResource(resource.getWorkspaceId(), resource.getResourceId());
+  }
+
+  @Test
+  public void duplicateControlledBucketNameRejected() {
+    final UUID workspaceId1 = createGcpWorkspace();
+    final ControlledGcsBucketResource resource1 =
+        ControlledResourceFixtures.makeDefaultControlledGcsBucketResource()
+            .workspaceId(workspaceId1)
+            .bucketName(BUCKET_NAME)
+            .build();
+
+    resourceDao.createControlledResource(resource1);
+
+    final UUID workspaceId2 = createGcpWorkspace();
+    final ControlledGcsBucketResource resource2 =
+        ControlledResourceFixtures.makeDefaultControlledGcsBucketResource()
+            .workspaceId(workspaceId2)
+            .name("another-bucket-resource")
+            .bucketName(BUCKET_NAME)
+            .build();
+
+    assertThrows(
+        DuplicateResourceException.class, () -> resourceDao.createControlledResource(resource2));
+
+    // clean up
+    resourceDao.deleteResource(resource1.getWorkspaceId(), resource1.getResourceId());
+    resourceDao.deleteResource(resource2.getWorkspaceId(), resource2.getResourceId());
+  }
+
+  // AI Notebooks are unique on the tuple {instanceId, location, projectId } in addition
+  // to the underlying requirement that resource ID and resource names are unique within a
+  // workspace.
+  @Test
+  public void duplicateNotebookIsRejected() {
+    final UUID workspaceId1 = createGcpWorkspace();
+    final ControlledResource resource1 =
+        ControlledResourceFixtures.makeDefaultAiNotebookInstance()
+            .workspaceId(workspaceId1)
+            .build();
+    resourceDao.createControlledResource(resource1);
+    assertEquals(
+        resource1, resourceDao.getResource(resource1.getWorkspaceId(), resource1.getResourceId()));
+
+    final ControlledResource resource2 =
```

Comment:
consider more meaningful names since we have a few ControlledResources here. maybeinstead of resource1-4  `initialResource`, `duplicatingResource`, `differentWorkspaceId`, and `differentLocation`,